Feature:  Implement tests streamers api

CeNTePeDe opened the pull request.

CeNTePeDe committed:
removed list_streams
add constant variables
created:
- get_data_stream
Pydentic models:
- StreamerIn
- StreamerOut

CeNTePeDe committed:
created:
- router to streamer
- exception(InvalidIdInputError)
- created DAO

CeNTePeDe committed:
added
- id to url
- added Exception
removed const variables
removed url
removed:
- InvalidIdInputError

CeNTePeDe committed:
added
- added exception

CeNTePeDe committed:
created:
- factories.py for streamers
- test_api_streamers.py
added:
- constant url for tests
- fixture mongo_mock

slavadan commented on database/connection_streamers.py:
```diff
@@ -0,0 +1,58 @@
+        streamer_update = self.collection.update_one(
+            {"id": id}, {"$set": streamer_data.dict()}
+        )
+        if streamer_update.modified_count == 0:
+            return None
+        return streamer_update.modified_count
```
просто возвращай streamer_update.modified_count

slavadan commented on tests/factories.py:
```diff
@@ -0,0 +1,29 @@
+    class Meta:
+        model = StreamerIn
+
+    id = "1111111"
```
бред id не может всегда быть 11111111, а если сделать через фарбрику еще 1 объект получится что 2 тебя два объекта с 111111111111111111111111 а id это id он должен быть уникальным

CeNTePeDe committed:
add function get_db()
rename fixture
add assert

created fixture:
- db
- streamer_build
created test for dao

slavadan commented on conftest.py:
```diff
@@ -0,0 +1,50 @@
+
+
+@pytest.fixture()
+def dao():
```
просто дао? слишком абстрактно

slavadan commented on conftest.py:
```diff
@@ -0,0 +1,50 @@
+
+@pytest.fixture()
+def streamer_build():
+    def car(**kwargs):
```
car???!! ты чо с шоурума стырила?

slavadan commented on conftest.py:
```diff
@@ -0,0 +1,50 @@
+
+@pytest.fixture()
+def streamer():
+    streamer = StreamerInFactory.stub().__dict__
```
не поправила

slavadan commented on tests/test_api_streamers.py:
```diff
@@ -0,0 +1,56 @@
+
+    response = client.post(url=URL_STREAMERS, json=streamer.dict())
+
+    assert response.status_code == 201
```
Slava Dan, [13. Jul 2023 at 20:16:24]:
когда тестишь пост запрос можно слать гет запрос чтобы проверить чтобы пост запрос корректно выполнился

или сразу запрос в бд слать

slavadan commented on conftest.py:
```diff
@@ -0,0 +1,51 @@
+@pytest.fixture()
+def dao_streamer_test():
+    mock_collection = mock.MagicMock(Collection)
+    dao = StreamerDAO()
+    dao.collection = mock_collection
```
StreamerDAO(collection)

slavadan commented on core/constant_variables.py:
```diff
@@ -4,12 +4,21 @@
 
-client: MongoClient = MongoClient("mongodb://user:1111@mongo:27017")
-db = client["product_db"]
+def get_db(db_name="product_db"):
+    client = MongoClient("mongodb://user:1111@mongo:27017")
+    db = client.get_database(db_name)
+    return db
```
убрать из констант

slavadan commented on database/connection_products.py:
```diff
@@ -4,11 +4,13 @@
 
 logger = logging.getLogger(__name__)
 
+db = get_db()
```
брать из настроек

slavadan commented on database/connection_streamers.py:
```diff
@@ -0,0 +1,60 @@
+
+logger = logging.getLogger(__name__)
+
+db = get_db()
```
брать из настроек

slavadan commented on parsers/parse_streamer.py:
```diff
@@ -0,0 +1,20 @@
+
+
+def get_data_stream(offset: Optional[int] = None) -> list[dict]:
+    from main import settings
```
пофиксить

slavadan commented on tests/test_api_streamers.py:
```diff
@@ -0,0 +1,62 @@
+    response = client.get(url=URL_STREAMERS + f"{id}")
+
+    assert response.status_code == 200
+    assert response.json()["id"] == "id_0"
```
== id

slavadan commented on tests/test_api_streamers.py:
```diff
@@ -0,0 +1,62 @@
+
+def test_get_streamer(db, client, streamer):
+    id = "id_0"
+    logger.info(f"id streamer {id} ")
```
убрать лог из тестов

slavadan commented on tests/test_api_streamers.py:
```diff
@@ -0,0 +1,62 @@
+def test_update_streamer(db, streamer, streamer_build, client):
+    new_streamer = streamer_build()
+    id = "id_0"
+    response = client.patch(url=URL_STREAMERS + f"{id}", json=new_streamer.dict())
```
плохой тест, который все еще ничего не показывает. Смысл проверять статус коды если вдруг модель не обновилась

slavadan commented on database/connection_streamers.py:
```diff
@@ -0,0 +1,60 @@
+            return None
+        return StreamerIn(**streamer_data)
+
+    def create_item(self, streamer_data: StreamerIn):
+        streamer_dict = streamer_data.dict()
+        logger.info("create streamer")
+        streamer = self.collection.find_one({"id": streamer_dict["id"]})
+        if streamer is None:
+            new_streamer = self.collection.insert_one(streamer_dict)
+            logger.info(f"new_streamer {new_streamer}")
+            return new_streamer
+
+        updated_streamer = self.update_item(streamer_dict["id"], streamer_data)
+        logger.info(f"updated streamer {updated_streamer}")
+        return updated_streamer
```
если метод называется create то он и должен делать create и возвращать объект созданного стримера. + нет тайпхинта

slavadan commented on tests/test_api_streamers.py:
```diff
@@ -0,0 +1,62 @@
+dao_streamer = StreamerDAO()
+
+
+def test_get_streamers(client):
+    response = client.get(url=URL_STREAMERS)
+
+    assert response.status_code == 200
+    assert len(response.json()) == len(dao_streamer.get_all_items())
```
длинну проверять конечно хорошо но было бы неплохо еще проверять ответ

slavadan commented on tests/test_api_streamers.py:
```diff
@@ -0,0 +1,62 @@
+    assert response.status_code == 200
+
+
+def test_create_streamer(db, client, streamer):
+    response = client.post(url=URL_STREAMERS, json=streamer)
+
+    assert response.json() == 1
+    assert response.status_code == 201
```
почему == 1. я ж говорил что ты должна проверять, что приходит. Вот ты создаешь стримера и метод тебе должен вернуть ответ со стримером, проверяй на одинаковые значения которые ты вставляла и получила

slavadan commented on tests/test_dao_streamers.py:
```diff
@@ -0,0 +1,39 @@
+    assert data_streamer_dict["user_id"] == data["user_id"]
+
+
+def test_update_item(dao_streamer_test, streamer, streamer_build):
+    item_id = streamer["id"]
+    update_data = streamer_build()
+    dao_streamer_test.collection.update_one.return_value.modified_count = 1
+    count = dao_streamer_test.update_item(item_id, update_data)
+
+    assert count == 1
```
во первых не по AAA. Во вторых мокать единичку не надо. Ты же тестишь логику обновления так зачем ее мокать. Смысл тогда от юнит тестов + было бы неплохо потом получить обновленную модели и проверить обновленные поля(я ж вроде в тг это описывал)

slavadan commented on tests/test_dao_streamers.py:
```diff
@@ -0,0 +1,39 @@
+    assert result == streamer
+
+
+def test_create_item(dao_streamer_test, streamer_build, streamer):
+    data_streamer = streamer_build()
+    data_streamer_dict = data_streamer.dict()
+    dao_streamer_test.collection.find_one.return_value = data_streamer_dict
+
+    data = dao_streamer_test.create_item(data_streamer)
+
+    assert data_streamer_dict["id"] == data["id"]
+    assert data_streamer_dict["user_id"] == data["user_id"]
```
зачемы ты мокаешь ответ. Ты же проверяешь логику создания

slavadan commented on tests/test_dao_streamers.py:
```diff
@@ -0,0 +1,39 @@
+    assert count == 1
+
+
+def test_delete_item(dao_streamer_test, streamer):
+    item_id = streamer["id"]
+    dao_streamer_test.collection.delete_one.return_value.deleted_count = 1
+    result = dao_streamer_test.delete_item(item_id)
+    dao_streamer_test.collection.delete_one.assert_called_once_with({"id": item_id})
+
+    assert result == 1
```
Не по AAA. опять мокаешь ответ. Мокать надо что-то постороннее

slavadan commented on conftest.py:
```diff
@@ -0,0 +1,51 @@
+    def test_db():
+        return db
+
+    monkeypatch.setattr("core.constant_variables.get_db", test_db)
```
delete
